File: backend/models/edit_params.py
from pydantic import BaseModel, Field


class EditParams(BaseModel):
    # Light
    exposure: float = Field(0.0, ge=-5.0, le=5.0)
    contrast: float = Field(0.0, ge=-100.0, le=100.0)
    highlights: float = Field(0.0, ge=-100.0, le=100.0)
    shadows: float = Field(0.0, ge=-100.0, le=100.0)
    whites: float = Field(0.0, ge=-100.0, le=100.0)
    blacks: float = Field(0.0, ge=-100.0, le=100.0)
    # Color
    temperature: float = Field(0.0, ge=-100.0, le=100.0)
    tint: float = Field(0.0, ge=-100.0, le=100.0)
    saturation: float = Field(0.0, ge=-100.0, le=100.0)
    vibrance: float = Field(0.0, ge=-100.0, le=100.0)
    # Detail
    sharpening: float = Field(0.0, ge=0.0, le=150.0)
    sharpening_radius: float = Field(1.0, ge=0.5, le=3.0)
    sharpening_detail: float = Field(25.0, ge=0.0, le=100.0)
    noise_reduction: float = Field(0.0, ge=0.0, le=100.0)
    noise_reduction_detail: float = Field(50.0, ge=0.0, le=100.0)
    noise_reduction_color: float = Field(25.0, ge=0.0, le=100.0)

File: swift-app/Sources/Doublenegative/Models/EditParams.swift
import Foundation

/// All non-destructive edit parameters for a single image.
/// Stored as JSON in SwiftData and used as the recipe format.
struct EditParams: Codable, Equatable, Hashable {
    // MARK: Light
    var exposure: Double    = 0   // -5.0 to +5.0 (EV stops)
    var contrast: Double    = 0   // -100 to +100
    var highlights: Double  = 0   // -100 to +100
    var shadows: Double     = 0   // -100 to +100
    var whites: Double      = 0   // -100 to +100
    var blacks: Double      = 0   // -100 to +100

    // MARK: Color
    var temperature: Double = 0   // -100 (cool) to +100 (warm)
    var tint: Double        = 0   // -100 (green) to +100 (magenta)
    var saturation: Double  = 0   // -100 to +100
    var vibrance: Double    = 0   // -100 to +100

    // MARK: Detail
    var sharpening: Double          = 0    // 0 to 150
    var sharpeningRadius: Double    = 1.0  // 0.5 to 3.0
    var sharpeningDetail: Double    = 25   // 0 to 100
    var noiseReduction: Double      = 0    // 0 to 100
    var noiseReductionDetail: Double = 50  // 0 to 100
    var noiseReductionColor: Double  = 25  // 0 to 100

    static let `default` = EditParams()
}
